enhydris/api/views.py

```
@@ -183,17 +184,62 @@ def bottom(self, request, pk=None, *, station_id):
         response.write(ts.get_last_record_as_string())
         return response
 
-    def _get_data(self, request, pk, format=None):
-        timeseries = get_object_or_404(models.Timeseries, pk=int(pk))
+    @action(detail=True, methods=["get"])
+    def chart(self, request, pk=None, *, station_id):
+        timeseries = get_object_or_404(models.Timeseries, pk=pk)
         self.check_object_permissions(request, timeseries)
+        start_date, end_date = self._get_date_bounds(request, timeseries)
+        # Drop rows with value "NaN"
+        data_frame = timeseries.get_data(
+            start_date=start_date, end_date=end_date
+        ).data.dropna(subset=["value"])
+        chart_data = self._sample_equidistant(data_frame)
+        return Response(
+            serializers.TimeseriesRecordChartSerializer(chart_data, many=True).data
+        )
 
+    def _sample_equidistant(self, df):
+        """
+        Divides dataframe/timeseries by time into "number_of_samples" equally,
+        and returns a list of dicts, where each dict contains timestamp, value
+        """
+        number_of_samples = min(
+            settings.ENHYDRIS_TS_GRAPH_BIG_STEP_DENOMINATOR, len(df.index)
+        )
+        min_time = df.index.min()
+        # To accommodate the first data point
+        current_time = min_time
+        interval = (df.index.max() - df.index.min()) / (number_of_samples - 1)
+        tolerance = interval / 2
+        result = []
+        for _ in range(number_of_samples):
+            try:
+                idx = df.index.get_loc(
+                    current_time, method="nearest", tolerance=tolerance
+                )
+                value = df.iloc[idx].value
+            except KeyError:
+                # If no matching index is found (ie exceeds tolerance)
+                # a KeyError is raised rather than a graceful null return
```

Please write comments with the same care as code. Use "i.e." instead of "ie" and make sure the syntax is correct, e.g. "the tolerance is exceeded" instead of "exceeds tolerance" and "returning null" instead of "graceful null return".

---

It's not actually wrong (of course `ie` is a mistake). I actually try to make the comments as short as possible (while obviously keeping the core words that explain it), and in "passive voice" whenever applicable.
So basically even after you pointed it out, I don't see the flaw with it to be honest. Anyway, I have changed them accordingly, just thought I would share my point of view :grin:

---

The flaws were
(1) "exceeds tolerance" What is the subject of "exceeds"?
(2) "a graceful null return": A graceful null return isn't "raised", so you can't really say "a KeyError is raised rather than a graceful null return", you need to say "... than a graceful null is returned" or something like that.